### PR TITLE
Upgrade tools package to 3.4

### PR DESCRIPTION
**Pull request by balinterdi**

### CODE REVIEW

**Comment by mansona** [comment]
@balinterdi can you verify that this has been solved with https://github.com/cardstack/cardstack/pull/453 ? 

Essentially ember-elsewhere had a bug that meant it wasn't compatible with Ember 3.3+

**Comment by balinterdi** [comment]
@mansona It would be amazing if that was solved. I'll try to give it a try today and report back.

**Comment by balinterdi** [comment]
It did solve it, thank you, that's fantastic!

I'll now weed out the 279 template lint errors and then it'll be ready to merge.

**Comment by balinterdi** [comment]
@ef4 Ok, so finally, after many merges from master (since all templates are linted now, every time an invalid template is merged to master, I fix it here :) ) this is ready to merge.

File: packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs [hunk 1]
```diff
@@ -1,8 +1,13 @@
 {{#if session.isAuthenticated}}
   {{#if network.iconUrl}}
     <div class="cardstack-rail-item cardstack-non-interactive">
-      <div class="cardstack-network-icon"
-           style={{css-url "background-image" network.iconUrl}} >{{network.name}}</div>
+      {{! template-lint-disable no-inline-styles }}
+      <div
+        class="cardstack-network-icon"
+        style={{css-url "background-image" network.iconUrl}}
+      >
```
Comment (Turbo87) [inline packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: Maybe we should just disable `no-inline-styles` globally? Doesn't seem like a particularly useful rule to me 🤔

File: packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs [hunk 2]
```diff
@@ -14,7 +19,7 @@
   {{/if}}
 
   {{#if (is-component "cardstack-workflow")}}
-    <div class="cardstack-rail-item {{if isOpen 'active'}}">
+    <div class={{concat "cardstack-rail-item " (if isOpen "active")}}>
```
Comment (Turbo87) [inline packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: why did you need `concat` here? this generally should work fine without it

File: packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs (hunk 1 above)
Comment (balinterdi) [reply packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: Good idea.

File: packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs (hunk 2 above)
Comment (balinterdi) [reply packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: Because we can only have double quotes in templates (a linting rule) and I can't embed a double-quoted string inside class="...".

Comment (Turbo87) [reply packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: hbs```
<div class="cardstack-rail-item {{if isOpen "active"}}">
```
does ☝️ not work?

**Comment by Turbo87** [comment]
> The template lint changes also relate to the upgrade, though, as the feature was introduced in 3.4, so the latter PR would have to have an extra commit that disables all template rules.

there are several ways to avoid that PR dependency:

- not include `ember-cli-template-lint` in the Ember 3.4 upgrade PR
- introduce `ember-cli-template-lint` first, without updating Ember

the introduction of template linting in general has nothing to do with the Ember update, it just happens to look that way because the blueprint now does it by default

File: packages/tools/package.json [hunk 3]
```diff
@@ -48,34 +50,34 @@
...
+    "eslint": "^5.9.0",
+    "eslint-plugin-ember": "^5.2.0",
+    "eslint-plugin-node": "^7.0.1",
+    "loader.js": "^4.7.0",
     "qunit-dom": "^0.8.0",
     "squishable-container": "^0.1.0"
   },
   "engines": {
-    "node": "^4.5 || 6.* || >= 7.*"
+    "node": "6.* || 8.* || >= 10.*"
```
Comment (Turbo87) [inline packages/tools/package.json]: since we only run CI on Node.js 8 we should consider adjusting this to `8.* || >= 10.*`

File: packages/tools/tests/dummy/app/templates/components/cardstack/post-isolated.hbs [hunk 4]
```diff
@@ -18,8 +18,13 @@
...
       <li>
-        {{cs-field comment.poster 'name'}}: {{cs-field comment 'body'}}
-        {{#cs-field-group comment name="karma" caption="Karma" fields="karmaValue,karmaType" as |ck|}}
+        {{cs-field comment.poster "name"}}: {{cs-field comment "body"}}
+        {{#cs-field-group
+          comment
+          name="karma"
+          caption="Karma"
+          fields="karmaValue,karmaType"
+        as |ck|}}
```
Comment (Turbo87) [inline packages/tools/tests/dummy/app/templates/components/cardstack/post-isolated.hbs]: seems unrelated?

File: packages/tools/addon/templates/components/cs-field-editor.hbs [hunk 5]
```diff
@@ -5,7 +5,9 @@ as |editorName editorOptions|
 }}
   {{#if (is-component editorName)}}
     {{#if firstError}}
-      <div class="field-editor--error-message">{{firstError}}</div>
+      <div class="field-editor--error-message" data-test-validation-error={{field}}>
```
[the remark this answers is not in the document]
Comment (balinterdi) [reply packages/tools/addon/templates/components/cs-field-editor.hbs]: Not strictly related. This is a fix for one of those tests that had passed miraculously in the past and break now. In this case, the unspecificity of a selector + a race condition made this test fail intermittently.

File: packages/tools/tests/dummy/app/templates/components/cardstack/post-isolated.hbs (hunk 4 above)
Comment (balinterdi) [reply packages/tools/tests/dummy/app/templates/components/cardstack/post-isolated.hbs]: It is. This was one of the merge conflicts I had to resolve manually during the interactive rebase and I decided to resolve with the version that also passes the template lint (probably that's also fixed in the template lint PR I'm submitting soon). This is just a "refactor", nothing was changed.

File: packages/tools/package.json (hunk 3 above)
Comment (balinterdi) [reply packages/tools/package.json]: Sorry, I overlooked this one, I'll create another issue.

File: packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs (hunk 2 above)
Comment (balinterdi) [reply packages/authentication/addon/templates/components/cardstack-authentication-launcher.hbs]: I'll try that in the linter PR, thank you.
